library/psa_crypto.c

```diff
@@ -2840,6 +2842,264 @@ psa_status_t psa_mac_verify_finish( psa_mac_operation_t *operation,
 /* Asymmetric cryptography */
 /****************************************************************/
 
+typedef enum
+{
+    PSA_SIGN_INVALID = 0,
+    PSA_SIGN_HASH = 1,
+    PSA_SIGN_MESSAGE
+} psa_sign_operation_t;
+
+typedef enum
+{
+    PSA_VERIFY_INVALID = 0,
+    PSA_VERIFY_HASH = 1,
+    PSA_VERIFY_MESSAGE
+} psa_verify_operation_t;
+
+static psa_status_t psa_sign_internal( mbedtls_svc_key_id_t key,
+                                       psa_sign_operation_t operation,
+                                       psa_algorithm_t alg,
+                                       const uint8_t * input,
+                                       size_t input_length,
+                                       uint8_t * signature,
+                                       size_t signature_size,
+                                       size_t * signature_length )
+{
+    psa_status_t status = PSA_ERROR_CORRUPTION_DETECTED;
+    psa_status_t unlock_status = PSA_ERROR_CORRUPTION_DETECTED;
+    psa_key_slot_t *slot;
+
+    *signature_length = 0;
+
+    if( operation == PSA_SIGN_INVALID )
+        return( PSA_ERROR_INVALID_ARGUMENT );
+    else
+    {
+        if( ! PSA_ALG_IS_SIGN_MESSAGE( alg ) )
```

Comment:
I'm still not quite friends with the way this logic is set up...

Correct me if I'm wrong, but the way I understand it, the logic would be:

- If operation is 'sign hash', only allow sign-hash algorithms, i.e. 'PSA_ALG_IS_SIGN_HASH( alg )`. Report an error otherwise.
- If operation is 'sign message', allow both sign-hash and sign-message algorithms, i.e. `PSA_ALG_IS_SIGN_MESSAGE( alg ) || PSA_ALG_IS_SIGN_HASH( alg )`. But since the definition of IS_SIGN_MESSAGE already accounts for the inclusion of IS_SIGN_HASH algorithms (which is honestly a bit counter-intuitive to me), that collapses to only `PSA_ALG_IS_SIGN_MESSAGE`.

And orthogonally to that, we need to verify that a signing algorithm which builds on a hash actually has a hash identifier encoded in its algorithm ID.

I believe this block (checking of algorithm value) could be written more clearly, and might even benefit from being structured as a switch-case on `operation`, since the operation type is an enum (meaning we can exhaustively address all potential values of the enum). It would also be more clear which kind of validation is performed where, since right now you're mixing validation of algorithm type v operation type, and internal algorithm type v. needed hash algorihm into a large blob.

Something like
```
switch(operation) {
  case sign_message:
    if(!PSA_ALG_IS_SIGN_MESSAGE(alg)) return error;
    break;
  case sign_hash:
    if(!PSA_ALG_IS_HASH_AND_SIGN(alg) return error;
    break;
  default:
    return error;
}

if (PSA_ALG_IS_HASH_AND_SIGN(alg) {
  // return error if algorithm ID doesn't specify which hash algo is used
}
```
would be cleaner (and more correct) IMO.